_posts/2018-11-06-mongodb-transactions.md

```diff
@@ -0,0 +1,157 @@
+
+A sidenote for those DBAs or developers who are already familiar with Oracle: the update structure in the WiredTiger cache is similar to how rollback segments and undo tablespaces work in Oracle, except that the snapshot information is kept in memory instead of on disk and you don't have to manage it.
+
+### Read your own writes
+MongoDB guarantees that you can read any writes you make inside your transaction, even before they are committed.
+It also guarantees that no other session can read your writes before they are committed.
+
+Again, these writes are handled by the snapshot structure inside the WiredTiger cache.
+
+### Write locks
+When two sessions are trying to update the same document at the same time, you get a write conflict.
+In MongoDB this conflict is handled by write locks.
+There are basically two conflict scenarios:
+
+- Before a *transaction* updates a document, it will try to acquire a write lock.
+If the document is already locked the transaction will fail.
+
+- Before a *non-transactional operation* tries to update a document, it will try to acquire a write lock.
+If the document is already locked, the operation will back off and retry until [MaxTimeMS](https://docs.mongodb.com/manual/reference/method/cursor.maxTimeMS/#cursor.maxTimeMS) is reached.
+
+Note that reads never block writes.
+MongoDB is also smart enough to recognize a so called no-op write: if the document you are trying to update was not changed by the update, it will not attempt to acquire a write lock.
+
+### Limitations
+Currently you can only use multi-document transactions with replica sets.
+Sharded clusters are not supported yet, though this feature is planned for a future release (4.1 perhaps?).
+
+Due to the WiredTiger cache pressure, long running transactions can be problematic, but the MongoDB developers are working on improving this for future releases and plan to support transactions running for several hours or even days.
+
+## Practical
+As for semantics, the MongoDB developers thankfully chose not to reinvent the wheel.
+Using transactions is similar to what most developers are used to on relational database systems.
+The precise syntax varies per programming language, so you will have to do some RTFM to learn it, but it always comes down to the following steps:
+
+1. Open session
+2. Start transaction
+3. Update multiple documents
+4. Commit or abort transaction
+
+For example, in Mongo Shell syntax a transaction typically looks like this:
+
+```javascript
+mySession = db.getMongo().startSession();
+mySession.startTransaction();
+mySession.getDatabase("mydb").coll1.insert({"foo" : "bar"});
+mySession.getDatabase("mydb").coll2.insert({"hello" : "world"});
+mySession.commitTransaction();
+```
+
+One important thing to note here is that once we open the session on the first line, every subsequent action must use the session variable ("mySession" in this case), otherwise they will be simple update operations not belonging to the transaction.
```

Comment:
```suggestion
One important thing to note here is that once we open the session on the first line, every subsequent action must use the session variable (`mySession` in this case), otherwise they will be simple update operations not belonging to the transaction.
```